File: Python/raw_threads.py
import threading
import multiprocessing
import time

from aux_tools import *
from clean_words import *

# ===============================

# Лише ASCII -- тому решта символів просто ігноруємо
def readData(file_name):
    # with open(file_name, 'r', encoding='utf-8') as datafile:
    with open(file_name, 'r', encoding='ascii', errors='ignore') as datafile:
        return datafile.read().split()


#! Винесено за межі "if __name__ == '__main__':"
# Щоб могти успадкувати WordsCount у кожному із субпроцесів
# Виглядає неохайно, зате просто.
config = read_config('data_input_conc.txt')

infile    = config["infile"]
out_by_a  = config["out_by_a"]
out_by_n  = config["out_by_n"]
threads_n = int(config["threads"])

etalon_a_file = config["etalon_a_file"]
# measurement_flags = int(config["measurement_flags"])

python_flags=config["python_flags"]
if 't' in python_flags:
    parent_class = threading.Thread
    lock_type = threading.Lock()
elif 'p' in python_flags:
    parent_class = multiprocessing.Process
    lock_type = multiprocessing.Lock()
    # manager = multiprocessing.Manager()
    # word_counter = manager.dict()
else:
    raise Exception('Neither threads no processes concurrency selected')

number_of_threads = threads_n


class WordsCount(parent_class):
    # lock = multiprocessing.Lock()
    # lock = threading.Lock()

    def __init__(self, words_list, word_counter, lock):
        super().__init__()
        self.words_list = words_list
        self.word_counter = word_counter
        self.lock = lock

    def run(self):
        local_dict = {}
        for word in self.words_list:
            word = cleanWord(word)
            if word not in local_dict:
                local_dict[word] = 1
            else:
                local_dict[word] += 1

        with self.lock:
            for i in local_dict.keys():
                if i in self.word_counter.keys():
                    self.word_counter[i] += local_dict[i]
                else:
                    self.word_counter[i] = local_dict[i]

if __name__ == '__main__':

    if 't' in python_flags:
        word_counter = {}
    elif 'p' in python_flags:
        manager = multiprocessing.Manager()
        word_counter = manager.dict()
    else:
        raise Exception('Neither threads no processes concurrency selected')

    start_time = (time.perf_counter(), time.process_time())

    input_list = readData(infile)

    start_anal_time = (time.perf_counter(), time.process_time())
    
    threads = []
    avg = len(input_list) / number_of_threads
    last = 0

    while last < len(input_list):
        threads.append(WordsCount(input_list[int(last):int(last + avg)], word_counter, lock_type))
        last += avg

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

    finish_time = (time.perf_counter(), time.process_time())

    print( 'Total:' )
    print( 'Wall time: {0}'.format(finish_time[0] - start_time[0]) )
    print( 'Process total time: {0}'.format(finish_time[1] - start_time[1]) )

    print( 'Analisys:' )
    print( 'Wall time: {0}'.format(start_anal_time[0] - start_time[0]) )
    print( 'Process total time: {0}'.format(start_anal_time[1] - start_time[1]) )

    write_sorted_by_key(word_counter,   out_by_a)
    write_sorted_by_value(word_counter, out_by_n)

    are_correct = True
    if etalon_a_file:
        are_correct = compareFiles(out_by_a, etalon_a_file);
    exit(not are_correct)

File: CPP/RawThreads.cpp
// This is a personal academic project. Dear PVS-Studio, please check it.
// PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com

#include <iostream>
#include <fstream>
#include <map>
#include <thread>
#include <vector>
#include <algorithm>
#include <set>
#include <mutex>
#include <chrono>
#include <atomic>
#include <sstream>

#include "aux_tools.hpp"
#include "clean_words.hpp"

#ifdef USE_UNORDERED_MAP
using map_type = std::unordered_map<std::string, unsigned int>;
#else // Default
using map_type = std::map<std::string, unsigned int>;
#endif

using namespace std;



vector<string> readData(ifstream& file) {
    vector<string> words;
    string word;
    while (file >> word) {
        words.push_back(word);
    }
    return words;
}


template<typename Iter>
void wordCounter(Iter start, Iter end, map_type& m, mutex& mtx)
{
    map_type localMp;

    for (auto i = start; i < end; ++i) {
        cleanWord(*i);
        ++localMp[*i];
    }
    lock_guard<mutex> lg(mtx);

    for(auto& item: localMp){
        m[item.first] += item.second;
    }
}


//! vec is not const, because latter it's iterators will be used to modify it's
//! elements -- see wordCounter.
vector<vector<string>::iterator> SplitVector(vector<string>& vec, unsigned n) {

    vector<vector<string>::iterator> outVec;
    auto part_length = vec.size() / n;
    auto sum = vec.begin();
    outVec.push_back(vec.begin());

    for(int i = 0; i<n-1;++i){
        advance(sum, part_length);
        outVec.push_back(sum);
        // cout<<outVec[i]<<endl;
    }
    outVec.push_back(vec.end());
    return outVec;
}


int main(int argc, char *argv[]) {
    auto config = read_config("data_input_conc.txt");

    string infile    = config["infile"];
    string out_by_a  = config["out_by_a"];
    string out_by_n  = config["out_by_n"];
    int    threads_n = str_to_val<unsigned>(config["threads"]);

    string etalon_a_file  = config["etalon_a_file"];

    auto start = get_current_time_fenced();

    ifstream data_file(infile);
    if (!data_file.is_open()) {
        cerr << "Error reading from file: " << infile << endl;
        return 1;
    }
    vector<string> words{readData(data_file)};
    data_file.close();

    auto readed = get_current_time_fenced();

    mutex mtx;
    map_type wordsMap;
    vector<thread> threads;
    auto work_parts{SplitVector(words, threads_n)};

    for (auto a = work_parts.begin(); a < work_parts.end()-1; ++a) {
        threads.emplace_back( wordCounter<vector<string>::iterator>,
                              *a, *(a + 1), ref(wordsMap), ref(mtx) );
    }

    for (auto& th : threads) th.join();

    auto finished = get_current_time_fenced();

    write_sorted_by_key(out_by_a, wordsMap);
    write_sorted_by_value(out_by_n, wordsMap);

    auto counting = finished - readed;
    auto total = finished - start;

    cout << "Total time    : " << to_us(total) << endl;
    cout << "Analisys time : " << to_us(counting) << endl;

    bool are_correct = true;
    if( !etalon_a_file.empty() )
    {
        are_correct = compareFiles(out_by_a, etalon_a_file);
    }
    return !are_correct;

}
